### ISSUE TITLE
Debugger step into with DynamicProxy?

### ISSUE THREAD

**Issue by andreas-marccain** [castleproject/Core#578]
Hi, 

we are using DynamicProxy to add some Logging to our ServiceCalls. 
Is there an easy way to enable the debugger "step into" feature in Visual Studio when using a proxy? 
By now the code in the interceptor and the target is skipped by the debugger.
A break point in the Interceptor or target get hit but we want to avoid this additional step. 
I had read the documentation but could not find any hints on this topic.

Thanks for your work on this project!

**Comment by stakx** [comment]
If I understand you correctly, it isn't that you want to step into DynamicProxy's source code; rather, you want to step into your interceptors and target objects from a proxy method call.

Generally speaking, it shouldn't be a problem to Step Into (F11) an interceptor, or into a target method. For example, given this code:

```csharp
         class Interceptor : IInterceptor
         {
             public void Intercept(IInvocation invocation)
/* 2: */     {
/* 3: */         invocation.Proceed();
             }
         }

         class Target : IFoo
         {
             public void Method()
/* 4: */     {
             }
         }

         var generator = new ProxyGenerator();
         var proxy = generator.CreateInterfaceProxyWithTarget<IFoo>(
             target: new Target(),
             interceptors: new Interceptor());
/* 1: */ proxy.Method();
```

If you stop execution at the line marked `1` and Step Into (F11), the debugger should reach the line marked `2`. Similarly, if you stop execution at `3` and Step Into again, the debugger should reach `4`. At least that's what happens on my machine, with a non-optimized Debug build, using Visual Studio 2019 (version 16.8.4).

I can imagine several things that might _possibly_ affect this &ndash; say, the interceptor class being located in a different project or in an external assembly; having the _Just My Code_ debugger setting enabled; making use of `[DebuggerNonUserCode]`, `[DebuggerStepThrough]`, etc. &ndash; you'll probably have to experiment a little.

Right now, I don't see anything that we'd need to change inside DynamicProxy to better enable this debugging scenario.

**Comment by andreas-marccain** [comment]
Hi,

I created a small sample under https://github.com/andreas-marccain/Castle.Core.DebugExample but have not managed to configure VS in the right way. Can u please check if the step into is working for you and share your debug settings?

My actual settings are:
![image](https://user-images.githubusercontent.com/72246779/108511499-4ad9b100-72c0-11eb-8c9d-675ccd298d39.png)


Thanks for your help!

**Comment by stakx** [comment]
If you think it'll help, my debugger settings are reproduced below. (I've deactivated quite a few options for performance reasons, since I'm working on an older notebook.)

![image](https://user-images.githubusercontent.com/104481/108532311-02c68880-72d8-11eb-95c4-e7f97a41d1d1.png)

**Comment by andreas-marccain** [comment]
And you are able to step into the interceptor or target in my example with these settings?

If I use the same settings the debugger still steps over the proxy.Method(); call.

**Comment by stakx** [comment]
Yes, debugging and stepping into the interceptor and target method (of your test solution) works fine with my installation.

**Comment by andreas-marccain** [comment]
Ok I figured out that the step into behavior only works if the "Enable Just My Code"  option is set.
Because of the interceptor and the target are located in other assemblies consumed by nuget packages we have to disable this option and then visual studio will no longer step into the target without setting a break point in the target.

Our nuget packages provide their symbols via source link but by now there is no option in vs to set the code of a nugetpackage as user code. It actually only depends on if .pdb files available or not. 

Thanks for your help on investigating this. 

We have solved this by controlling the presence of an interceptor with a flag for debugging needs on the target. But this is just an ugly workaround.
